crates/bevy_ecs/src/schedule/executor/single_threaded.rs

```diff
@@ -96,16 +96,26 @@ impl SystemExecutor for SingleThreadedExecutor {
             let system = &mut schedule.systems[system_index];
             if is_apply_deferred(system) {
                 self.apply_deferred(schedule, world);
-            } else {
-                let res = std::panic::catch_unwind(AssertUnwindSafe(|| {
+                continue;
+            }
+
+            let res = std::panic::catch_unwind(AssertUnwindSafe(|| {
+                if system.is_exclusive() {
                     system.run((), world);
-                }));
-                if let Err(payload) = res {
-                    eprintln!("Encountered a panic in system `{}`!", &*system.name());
-                    std::panic::resume_unwind(payload);
+                } else {
+                    // Use run_unsafe to avoid immediately applying deferred buffers
+                    let world = world.as_unsafe_world_cell();
```

Comment:
I'd rather we just make `single_threaded` executor work the same way as `simple` executor. We should make it clear that `apply_deferred` is just an optimization to allow systems to run in parallel. Any single-threaded executor should apply commands immediately -- it has access to `&mut World` so there's little reason not to.
Although I suppose deprecating `SimpleExecutor` can be for another PR.